File: lv1_module2/ros2_ws/src/turtle_cpp/src/distance_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float32.hpp>
#include <turtlesim/msg/pose.hpp>
#include <cmath>

class DistancePublisher : public rclcpp::Node
{
public:
    DistancePublisher() : Node("distance_publisher_cpp")
    {
        pose_sub_ = this->create_subscription<turtlesim::msg::Pose>(
            "/turtle1/pose", 10,
            [this](const turtlesim::msg::Pose::SharedPtr msg) {
                last_pose_ = *msg;
                pose_received_ = true;
            });

        dist_pub_ = this->create_publisher<std_msgs::msg::Float32>(
            "/turtle_distance", 10);

        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(&DistancePublisher::timer_callback, this));

        RCLCPP_INFO(this->get_logger(), "C++ Distance publisher started at 10 Hz");
    }

private:
    void timer_callback()
    {
        if (!pose_received_) return;

        auto msg = std_msgs::msg::Float32();
        msg.data = std::sqrt(
            last_pose_.x * last_pose_.x + last_pose_.y * last_pose_.y);
        dist_pub_->publish(msg);
    }

    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_sub_;
    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr dist_pub_;
    rclcpp::TimerBase::SharedPtr timer_;
    turtlesim::msg::Pose last_pose_;
    bool pose_received_ = false;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<DistancePublisher>());
    rclcpp::shutdown();
    return 0;
}

File: lv1_module2/ros2_ws/src/turtle_py/turtle_py/distance_publisher.py
import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose
from std_msgs.msg import Float32
import math


class DistancePublisher(Node):
    def __init__(self):
        super().__init__('distance_publisher')

        self.declare_parameter('publish_rate', 10.0)
        self.last_pose = None

        self.pose_sub = self.create_subscription(
            Pose, '/turtle1/pose', self.pose_callback, 10)
        self.dist_pub = self.create_publisher(Float32, '/turtle_distance', 10)

        rate = self.get_parameter('publish_rate').value

        # 예외 처리: publish_rate가 0 이하이면 경고 후 기본값 사용
        if rate <= 0:
            self.get_logger().warn(
                f'Invalid publish_rate={rate}, using default 10.0 Hz')
            rate = 10.0

        timer_period = 1.0 / rate
        self.timer = self.create_timer(timer_period, self.timer_callback)

        self.get_logger().info(
            f'Distance publisher started at {rate} Hz')

    def pose_callback(self, msg):
        """구독 콜백: 최신 pose만 저장 (발행은 여기서 안 함!)"""
        self.last_pose = msg

    def timer_callback(self):
        """타이머 콜백: 저장된 pose로 거리 계산 후 발행"""
        if self.last_pose is None:
            return  

        distance = math.sqrt(
            self.last_pose.x ** 2 + self.last_pose.y ** 2)

        msg = Float32()
        msg.data = distance
        self.dist_pub.publish(msg)


def main(args=None):
    rclpy.init(args=args)               
    node = DistancePublisher()      
    try:
        rclpy.spin(node)              
    except KeyboardInterrupt:
        pass                        
    finally:
        node.destroy_node()
        try:
            rclpy.shutdown()
        except Exception:
            pass


if __name__ == '__main__':
    main()